profiling/src/collections/identifiable/mod.rs

```diff
@@ -66,13 +67,34 @@ pub trait Dedup<T: Item> {
     /// Panics if the number of items overflows the storage capabilities of
     /// the associated Id type.
     fn dedup(&mut self, item: T) -> <T as Item>::Id;
+
+    /// Deduplicate the Item, and check if the generated Id is valid.
+    fn checked_dedup(&mut self, item: T) -> anyhow::Result<<T as Item>::Id>;
 }
 
 impl<T: Item> Dedup<T> for FxIndexSet<T> {
     fn dedup(&mut self, item: T) -> <T as Item>::Id {
         let (id, _) = self.insert_full(item);
         <T as Item>::Id::from_offset(id)
     }
+
+    /// In incident 35390 (JIRA PROF-11456) we observed invalid location_ids being present in
+    /// emitted profiles. It's not likely that the incorrect ids are coming from the underlying
+    /// collection, but we're doing extra checks here so that if we see incorrect ids again,
+    /// we are 100% sure they were not introduced at this stage.
+    fn checked_dedup(&mut self, item: T) -> anyhow::Result<<T as Item>::Id> {
+        let (id, _) = self.insert_full(item);
+
+        anyhow::ensure!(
+            id < self.len(),
+            "out of bounds id generated {:?}, len was {:?}",
+            id,
+            self.len()
+        );
+        small_non_zero_pprof_id(id).ok_or_else(|| anyhow!("invalid id generated {:?}", id))?;
```

Comment:
Can use `.with_context()` here.

Reply:
Fixed in 2ddd704d9e30798ac219c34673ea71e7c41baf64